### ISSUE TITLE
[Bug]: matmul op calculation results are wrong when input size exceeds 2048 on GPU

### ISSUE THREAD

**Issue by ecneil** [openvinotoolkit/openvino#22613]
### OpenVINO Version

2023.0-2023.3

### Operating System

Ubuntu 22.04 (LTS)

### Device used for inference

GPU

### Framework

None

### Model used

None

### Issue description

We encountered an output problem with one of our models.
After isolating the problem, we found that if the input is greater than 2048, the matmul operation returns incorrect output.
Also, this problem did not occur prior to 2023.0.

### Step-by-step reproduction

Execute the following code.

``` python
import openvino.runtime as ov
import numpy as np

def create_matmul_model(size):
    input_data1 = ov.opset10.parameter([1, 1, size], ov.Type.f32)
    input_data2 = ov.opset10.parameter([1, size, 1], ov.Type.f32)
    output = ov.opset10.matmul(data_a=input_data1,
                                data_b=input_data2,
                                transpose_a=False,
                                transpose_b=False)
    model = ov.Model([output], [input_data1, input_data2], "model")
    return model

core = ov.Core()

input_size = (2048, 2049)

for size in input_size:
    model = create_matmul_model(size)
    compiled_cpu_model = core.compile_model(model, "CPU")
    compiled_gpu_model = core.compile_model(model, "GPU")
    cpu_infer_request = compiled_cpu_model.create_infer_request()
    gpu_infer_request = compiled_gpu_model.create_infer_request()

    input_1_name = model.inputs[0].any_name
    input_2_name = model.inputs[1].any_name
    input_1 = np.ones((1, 1, size), np.float32)
    input_2 = np.ones((1, size, 1), np.float32)
    input_ = {input_1_name: input_1,
              input_2_name: input_2}

    cpu_pred = cpu_infer_request.infer(input_)
    gpu_pred = gpu_infer_request.infer(input_)
    cpu_result = list(cpu_pred.values())[0]
    gpu_result = list(gpu_pred.values())[0]
    eq = np.all(cpu_result == gpu_result)
    print(f"input size: {size}\n"
          f"output is {'same' if eq else 'different'}\n"
          f"cpu_result_max_value: {cpu_result.max()}\n"
          f"gpu_result_max_value: {gpu_result.max()}\n")
```

### Relevant log output

on 2023.3
```
input size: 2048
output is same
cpu_result_max_value: 2048.0
gpu_result_max_value: 2048.0

input size: 2049
output is different
cpu_result_max_value: 2049.0
gpu_result_max_value: 2048.0
```

on 2022.3.1
```
input size: 2048
output is same
cpu_result_max_value: 2048.0
gpu_result_max_value: 2048.0

input size: 2049
output is same
cpu_result_max_value: 2049.0
gpu_result_max_value: 2049.0
```

### Issue submission checklist

- [X] I'm reporting an issue. It's not a question.
- [X] I checked the problem with the documentation, FAQ, open issues, Stack Overflow, etc., and have not found a solution.
- [X] There is reproducer code and related data files such as images, videos, models, etc.

**Comment by geunhwan** [comment]
@ecneil Could you let me know HW SKU on which you're testing?

**Comment by ecneil** [comment]
Here I have confirmed the problem with the following on iGPUs on the CPUs below
- Core i5 1135G7
- Core i3 10100
- Core i5 13500

**Comment by ahnyoung-paul** [comment]
I found this wrong result after a [commit ](https://github.com/openvinotoolkit/openvino/commit/e23b8492c56126b838d2866553a93e490c2bb74d)that changed the default inference precision from undefined to fp16. When fp16 is set to the 
 default inference precision of the model, fp32 input data is converted to fp16 data internally and gemm calculates output with fp16 data. However, in the case of fp16 data type, input size 2049 has an accuracy issue because the mantissa of FP16 is only 10 bits. I think it makes the current wrong output issue.
 
 If you set INFERENCE_PRECISION_HINT with fp32 when the model compiling, you can get the right result.
 
`config = {'INFERENCE_PRECISION_HINT': 'FP32'}`
`compiled_gpu_model = core.compile_model(model, "GPU", config)`

**Comment by ahnyoung-paul** [comment]
This is my test results

------------------------------------
**default infer_precision : fp16 .........**
    GPU - Intel GPU plugin: 2024.1.2024.1.0-14597-ed6b3a53337

**Set infer_precision to f32**
input size: 1024
gpu_result_max_value: 1024.0

input size: 1025
gpu_result_max_value: 1025.0

input size: 1026
gpu_result_max_value: 1026.0

input size: 2046
gpu_result_max_value: 2046.0

input size: 2047
gpu_result_max_value: 2047.0

input size: 2048
gpu_result_max_value: 2048.0

input size: 2049
gpu_result_max_value: 2049.0

input size: 2050
gpu_result_max_value: 2050.0

input size: 3072
gpu_result_max_value: 3072.0

**Comment by ecneil** [comment]
Hi @ahnyoung-paul, Thank you for your investigation.

I did not think that the rounding error was the cause because the first model I confirmed the problem with was converted as fp16 by the model optimizer.
(I thought that when a model as fp16 was inferred on a GPU, all calculations were computed at fp16.)

Just to be sure, am I right in assuming that even when a model quantized with int8 is inferred on a GPU, the precision of calculations not involving quantization is done according to "INFERENCE_PRECISION_HINT" (or fp16 if unspecified)?

**Comment by yeonbok** [comment]
Hi @ecneil 

>> (I thought that when a model as fp16 was inferred on a GPU, all calculations were computed at fp16.)
Yes, it is fp16 and since the # of significant bits are just 11, so it is overflowed. 


>> even when a model quantized with int8 is inferred on a GPU, the precision of calculations not involving quantization is done according to "INFERENCE_PRECISION_HINT" (or fp16 if unspecified)?

No, i8 quantized layers are to be inferenced as i8 for quantized model.

**Comment by ecneil** [comment]
Hi @yeonbok 

I guess my question was not clear.

I know that when using POT with accuracy-aware quantization or NNCF with quantize_with_accuracy_control, there will still exist a non-quantized layer with floating point calculation in the int8 quantization model.
In this case, I think the precision of the non-quantized layer is calculated by "INFERENCE_PRECISION_HINT" (or fp16 if not specified when inferring on GPU), is that correct?
Or it depends on the original model's precision ?

**Comment by yeonbok** [comment]
Hi @ecneil 

The steps are like this 
1. Decide fp precision : fp16 if hint is not set, fp32 if hint is set for accuracy or precision hint fp32
2. ConvertPrecision with the fp precision as decided. However some precision sensitive layers might be remaining as fp32 though the hint is not set  (e.g., MVN) 
4. Then the FakeQuantize layers are expanded as much as possible by Low Precision Transform => i.e., i8 layers are decided (not only the initial layers in the POT result with FakeQuantize, but it explores more possibility)

So for your initial question 
- Basically fp layers are to be fp precision, default is fp16 and can be set as fp32 with the hint. Though the hint is not set some layers can be remaining with fp32 with the internal policy 
- The internal transform decides i8 layers, to be as many as possible, by its policy

**Comment by ecneil** [comment]
Hi @yeonbok, thank you for your answer.

I understand clearly how the choice of computational precision of layers on GPU.

Issue is now closed.
